jsonrpcbase/main.py

```diff
@@ -165,122 +150,93 @@ def call_py(self, jsondata):
+                        respond = self._handle_request(request_, metadata)
                     except JSONRPCError as e:
-                        respond = self._get_err(e,
-                                                request_['id'],
-                                                request_['jsonrpc'])
-
+                        respond = self._get_err(e, request_['id'], request_['jsonrpc'])
                     # Don't respond to notifications
                     if respond is not None:
                         responds.append(respond)
-
                 if responds:
                     return responds
-
                 # Nothing to respond.
                 return None
             else:
                 # empty dict, list or wrong type
                 raise InvalidRequestError
-
         except InvalidRequestError as e:
             return self._get_err(e, request['id'])
+        except InvalidIDType as err:
+            return self._invalid_id_type_response(err, request)
         except JSONRPCError as e:
             return self._get_err(e, request['id'], request['jsonrpc'])
+        except jsonschema.exceptions.ValidationError as e:
+            return self._invalid_params_response(e, request['id'], request['jsonrpc'])
+
+    def _invalid_id_type_response(self, err, request):
+        """
+        Response for an invalid `id` field type
+        """
+        err = {
+            'message': err.message,
+            'code': err.code,
+        }
+        resp = {'id': None, 'error': err}
+        return resp
+
+    def _invalid_params_response(self, err, id=None, jsonrpc=DEFAULT_JSONRPC):
+        """
+        Returns an error message for a jsonschema validation error on the params.
+        """
+        resp = {'id': id}
+        self._fill_ver(jsonrpc, resp)
+        resp['error'] = {
+            'message': err.message,
+            'code': -32602,
+        }
+        return resp
 
     def _get_err(self, e, id=None, jsonrpc=DEFAULT_JSONRPC):
         """
         Returns jsonrpc error message.
         """
         # Do not respond to notifications when the request is valid.
-        if not id and not isinstance(e, ParseError) and not isinstance(e, InvalidRequestError):
+        if id is None and not isinstance(e, ParseError) and not isinstance(e, InvalidRequestError):
             return None
-
         respond = {'id': id}
-
-        if isinstance(jsonrpc, int):
-            # v1.0 requires result to exist always.
-            # No error codes are defined in v1.0 so only use the message.
-            if jsonrpc == 10:
-                respond['result'] = None
-                respond['error'] = e.dumps()['message']
-            else:
-                self._fill_ver(jsonrpc, respond)
-                respond['error'] = e.dumps()
-        else:
-            respond['jsonrpc'] = jsonrpc
-            respond['error'] = e.dumps()
-
+        self._fill_ver(jsonrpc, respond)
+        respond['error'] = e.dumps()
```

Comment:
Discussed. JSON RPC 1.1 and 2.0 have the same error style. (1.0 support removed)